client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HeaderDecoder.java

```diff
@@ -1,79 +1,152 @@
+               }
+               status = codec.readHeader(in, operation.header(), channelFactory, ctx.channel().remoteAddress());
+               checkpoint(State.READ_PAYLOAD);
+            case READ_PAYLOAD:
+               if (trace) {
+                  log.tracef("Decoding payload for %s on %s", operation, ctx.channel());
+               }
+               operation.acceptResponse(in, status, this);
+               checkpoint(State.READ_MESSAGE_ID);
+               break;
+         }
+      } catch (Signal signal) {
+         throw signal;
+      } catch (Exception e) {
+         // If this is server error make sure to restart the state of decoder
+         checkpoint(State.READ_MESSAGE_ID);
+         throw e;
+      }
+   }
+
+   @Override
+   public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+      if (operation != null) {
+         operation.exceptionCaught(ctx, cause);
+      } else {
+         TransportException transportException = log.errorFromUnknownOperation(ctx.channel(), cause, ctx.channel().remoteAddress());
+         for (HotRodOperation<?> op: incomplete.values()) {
             try {
-               ctx.pipeline().remove(this);
-               ctx.pipeline().remove(operation);
-               operation.releaseChannel(ctx.channel());
-               timeoutFuture.cancel(false);
+               op.exceptionCaught(ctx, transportException);
             } catch (Throwable t) {
-               try {
-                  ctx.channel().close();
-               } catch (Throwable t2) {
-                  t.addSuppressed(t2);
-               }
-               operation.completeExceptionally(t);
-               return;
+               log.errorf(t, "Failed to complete %s", op);
             }
-            operation.complete(result);
-            break;
+         }
+         if (trace) {
+            log.tracef(cause, "Requesting %s close due to exception", ctx.channel());
+         }
+         ctx.close();
+      }
+   }
+
+   @Override
+   public void channelInactive(ChannelHandlerContext ctx) {
+      for (HotRodOperation<?> op: incomplete.values()) {
+         try {
+            op.channelInactive(ctx.channel());
+         } catch (Throwable t) {
+            log.errorf(t, "Failed to complete %s", op);
+         }
       }
    }
 
+   public CompletableFuture<Void> allCompleteFuture() {
+      return CompletableFuture.allOf(incomplete.values().toArray(new CompletableFuture[incomplete.size()]));
+   }
+
+   @Override
+   public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+      if (evt instanceof ChannelPoolCloseEvent) {
+         closing = true;
+         allCompleteFuture().whenComplete((nil, throwable) -> {
+            ctx.channel().close();
+         });
+      } else if (evt instanceof IdleStateEvent) {
+         // If we have incomplete operations this channel is not idle!
+         if (!incomplete.isEmpty()) {
+            return;
```

Comment:
Curious, should we possibly complete exceptionally the operations saying there was a timeout or something? Or this should never happen if there are any incomplete?

Reply:
No, we shouldn't do that; Idle timeout can be lower than socket timeout (which was defining how long the operation can take with synchronous operations and is used for timeout for each operation). If there wasn't this check the channel would be closed and the operations would complete exceptionally as you suggest, but we don't want to fail the operation just because the server is taking somewhat longer. There's even a test that lowers the idle timeout ridiculously.